lib/buildenv/builder.pl

```diff
@@ -0,0 +1,303 @@
+    }
+
+    # If target already exists and both targets resolves to the same path, skip
+    if (
+        defined $oldTarget && $oldTarget ne "" &&
+        defined abs_path($target) && defined abs_path($oldTarget) &&
+        abs_path($target) eq abs_path($oldTarget)
+    ) {
+        # Prefer the target that is not a symlink, if any
+        if (-l $oldTarget && ! -l $target) {
+            $symlinks{$relName} = [$target, $priority];
+        }
+        return;
+    }
+
+    # If target already exists as a symlink to a file (not a
+    # directory) in a higher-priority package, skip.
+    if (defined $oldTarget && $priority > $oldPriority && $oldTarget ne "" && ! -d $oldTarget) {
+        return;
+    }
+
+    # If target is supposed to be a directory but it isn't, die with an error message
+    # instead of attempting to recurse into it, only to fail then.
+    # This happens e.g. when pathsToLink contains a non-directory path.
+    if ($oldTarget eq "" && ! -d $target) {
+        die "not a directory: `$target'\n";
+    }
+
+    unless (-d $target && ($oldTarget eq "" || -d $oldTarget)) {
+        # Prepend "dangling symlink" to paths if applicable.
+        my $targetRef = prependDangling($target);
+        my $oldTargetRef = prependDangling($oldTarget);
+
+        if ($ignoreCollisions) {
+            warn "collision between $targetRef and $oldTargetRef\n" if $ignoreCollisions == 1;
+            return;
+        } elsif ($checkCollisionContents && checkCollision($oldTarget, $target)) {
+            return;
+        } else {
+            die "collision between $targetRef and $oldTargetRef\n";
+        }
+    }
+
+    findFilesInDir($relName, $oldTarget, $ignoreCollisions, $checkCollisionContents, $oldPriority) unless $oldTarget eq "";
+    findFilesInDir($relName, $target, $ignoreCollisions, $checkCollisionContents, $priority);
+
+    $symlinks{$relName} = ["", $priority]; # denotes directory
+}
+
+
+my %done;
+my %postponed;
+
+sub addPkg {
+    my ($pkgDir, $ignoreCollisions, $checkCollisionContents, $priority)  = @_;
+
+    return if (defined $done{$pkgDir});
+    $done{$pkgDir} = 1;
+
+    findFiles("", $pkgDir, "", $ignoreCollisions, $checkCollisionContents, $priority);
+
+    # <flox>
+    #
+    #   When rendering floxEnvs treat propagated-build-inputs as if they
+    #   were propagated-user-env-packages so that required packages already
+    #   present in the closure can be found from the one environment path.
+    #   This is particularly relevant for interpreted languages like python
+    #   that can then use a single value for PYTHONPATH rather than having
+    #   to rely upon walking setup hooks for constructing a long PYTHONPATH
+    #   during a potentially-unbounded instantiation.
+    #
+    # my $propagatedFN = "$pkgDir/nix-support/propagated-user-env-packages";
+    foreach my $propagatedFN (
+        "$pkgDir/nix-support/propagated-user-env-packages", "$pkgDir/nix-support/propagated-build-inputs"
+    ) {
```

Comment:
I just wanted to note that these lines are literally the only differences from `builder.pl` as found in nixpkgs upstream, and I've left the indentation of the following code block unchanged to make it easy to incorporate updates from upstream over time.

Reply:
I like splitting out a commit when I copy in a file so I can diff against that commit to see changes

Reply:
Discussed with Matthew offline that we added a README.flox to make clear how these files were created and updated over time.